### DISCUSSION TITLE
Add an event to intercept code in `Address::beforeValidate()`

### DISCUSSION THREAD

**Discussion by rob-c-baker** [craftcms/cms#11788] [Ideas] [answered]
I have a site I am in the process of moving to Craft / Commerce v4 so all addresses are migrated to the new `Address` element.

This is causing an issue as follows _(for our UK / GB based addresses)_:

The above method pulls address formatting from an address format repository: [commerceguys/addressing](https://github.com/commerceguys/addressing) which in turn pulls it's formatting data from a [Google-run repository](https://github.com/google/libaddressinput). An issue affecting us in `commerceguys/addressing`: https://github.com/commerceguys/addressing/issues/111 has been referred upstream here: https://github.com/google/libaddressinput/issues/193 and it got closed based on the idea that the Royal Mail say it's not essential to include the `administrativeArea` field in an address.

Our business would most definitely like to have `administrativeArea` (i.e. "county" in local terms) appearing on our addresses as it is often necessary for addresses to include this line (independent from the purpose of using an address as a postal destination).

Because our address data was migrated, we are in a situation where sometimes an `administrativeArea` appears in an address and sometimes is does not because the data was migrated, but depending whether validation is triggered or not, the code below sets our beloved field to null:

https://github.com/craftcms/cms/blob/520908d923bae2fafa3c151bd828cc4383672b03/src/elements/Address.php#L457-L478

A great solution to this and to avoid getting changes passed and released through 2 upstream repositories would be a new event, maybe `Address::BEFORE_VALIDATE` or similar that would let us developers prevent field(s) from being nulled if we need to. Is that possible?

**Comment by brandonkelly** [comment, marked as answer]
Just added three new methods and events to `craft\services\Addresses` for Craft 4.3: (804f8e817ba83c1f8101c882b054e4abec5a00eb)

- `getUsedFields()` + `EVENT_DEFINE_USED_FIELDS`
- `getUsedSubdivisionFields()` + `EVENT_DEFINE_USED_SUBDIVISION_FIELDS`
- `getFieldLabel()` + `EVENT_DEFINE_FIELD_LABEL`

These can be used to modify the list of fields that the Addressing library considers “used” for each country code, as well as the country code-specific label for them.

In your case, you could add the `administrativeAria` for the UK, and label it as “County”, with the following module code:

```php
use CommerceGuys\Addressing\AddressFormat\AddressField;
use craft\events\DefineAddressFieldLabelEvent;
use craft\events\DefineAddressFieldsEvent;
use craft\services\Addresses;
use yii\base\Event;

Event::on(
    Addresses::class,
    Addresses::EVENT_DEFINE_USED_FIELDS,
    function(DefineAddressFieldsEvent $event) {
        if ($event->countryCode === 'GB') {
            $event->fields[] = AddressField::ADMINISTRATIVE_AREA;
        }
    }
);

Event::on(
    Addresses::class,
    Addresses::EVENT_DEFINE_FIELD_LABEL,
    function(DefineAddressFieldLabelEvent $event) {
        if (
            $event->countryCode === 'GB' &&
            $event->field === AddressField::ADMINISTRATIVE_AREA
        ) {
            $event->label = 'County';
        }
    }
);
```

4.3 is still in active development, but if you want to get on it early for this, change your `craftcms/cms` constraint in composer.json to `4.3.x-dev as  4.3.0-alpha` and run `composer update`.

**Reply by rob-c-baker** [reply]
This looks perfect - many thanks. Aside from my use-case, this should enhance the flexibility of addresses significantly.

**Reply by brandonkelly** [reply]
Craft 4.3 was released earlier this week with those changes!

**Reply by DanielDarrenJones** [reply]
@brandonkelly Sorry for messaging on such an old thread, but is there a way to mark the County field here as required? For example the Administrative Area when United States is selected is required, but when assigning it to the UK through this event it doesn't show as required?

**Reply by brandonkelly** [reply]
@DanielDarrenJones It might be possible to add a `required` rule to `county` from `EVENT_DEFINE_RULES`. Otherwise you could manually check and add a validation error when not set, via `EVENT_AFTER_VALIDATE`.

**Comment by bymayo** [comment]
@brandonkelly Is there any way of doing this, but without it being a input field, it still displaying a list of Counties? It would be good if there was an event to extend that field due to this issue with UK Counties

**Reply by brandonkelly** [reply]
That’s now possible in Craft 4.5.0 (#13361).
